phantom-drift/introspection.proto

```diff
@@ -0,0 +1,351 @@
+    bool listening  = 2;
+    // whether this transport is enabled for dialing.
+    bool dialing    = 3;
+    // the multiaddrs this transport is listening on, if enabled for listening.
+    repeated string listen_multiaddrs = 4;
+}
+
+// Dialer encapsulates stats and state about the swarm dialer.
+message Dialer {
+    // InflightDial represents a dial that is in progress.
+    message InflightDial {
+        PeerId peer_id              = 1;
+        uint32 addrs_cnt            = 2;
+        ResultCounter attempts  = 3;
+    }
+
+    // the number of concurrent dials allowed.
+    uint32 throttle_max     = 1;
+    // the number of concurrent dials currently active.
+    uint32 throttle_curr    = 2;
+
+    // peer dial stats.
+    ResultCounter peer_dials    = 3;
+    // address dial stats.
+    ResultCounter addrs_dials   = 4;
+    // dial stats per transport.
+    map<string, ResultCounter> dials_per_transport = 5;
+
+    // listing of in-progress dials.
+    repeated InflightDial inflight = 6;
+}
+
+// Swarm reports stats and state of the swarm subsystem.
+message Swarm {
+    // SwarmCounterGroup models a group of counters; encapsulation is important to spare 1-byte field IDs (1-15).
+    message SwarmCounterGroup {
+        // incoming items opened.
+        TimeWindowCounter incoming_opened = 1;
+        // outgoing items opened.
+        TimeWindowCounter outgoing_opened = 2;
+        // items closed.
+        TimeWindowCounter closed = 3;
+        // current count (instantaneous reading).
+        uint32 active = 4;
+    }
+
+    // TaggedSwarmCounterGroup models a group of counters segregated by a tag.
+    // See docs for SwarmCounterGroup for inner fields.
+    message TaggedSwarmCounterGroup {
+        map<string, TimeWindowCounter> incoming_opened = 1;
+        map<string, TimeWindowCounter> outgoing_opened = 2;
+        map<string, TimeWindowCounter> closed = 3;
+        map<string, uint32> active = 4;
+    }
+
+    // connections statistics.
+    SwarmCounterGroup conn_stats = 1;
+    // stream statistics
+    SwarmCounterGroup stream_stats = 2;
+    // stream statistics per protocol.
+    TaggedSwarmCounterGroup stream_by_proto_stats = 3;
+
+    // connection listing.
+    repeated Connection conns = 4;
+    // streams listing.
+    repeated Stream streams = 5;
+
+    // inner subsystems
+    Dialer dialer = 6;
+
+    // We want to be frugal with the 1-15 field id range, as they are represented by 1 byte only.
+    reserved 7 to 15;
+}
+
+// Host is a top-level object conveying a cross-cutting view of the entire system, including all subsystems.
+// TODO: WIP right now we're only covering the Swarm subsystem.
+message Host {
+    // our peer id.
+    PeerId id = 1;
+    // the transports enabled in this host.
+    repeated Transport transports = 2;
+    // the protocols attached to this host.
+    repeated string protocols = 3;
+    // the swarm subsystem.
+    Swarm swarm = 4;
```

Comment:
What do you think about adding a `message Subsystems`, where we would include `Swarm`, as well as the remaining ones?

Reply:
Yes, probably more scalable that way ;-)